http/middleware/log_request.go

```diff
@@ -2,44 +2,167 @@ package middleware
+)
+
+// LogRequest logs the a LogRequestRecord using the provided handler.
 //
-// LogRequest scrubs the values for the following keys:
+// For the LogRequestRecord.URI, LogRequest masks query params matching these keys with trails.LogMaskVal:
 // - password
 //
-// if logger.Logger is nil, NoopAdapter returns and this middleware does nothing.
-func LogRequest(ls logger.Logger) Adapter {
-	if ls == nil {
+// If handler is nil, NoopAdapter returns and this middleware does nothing.
+func LogRequest(handler slog.Handler) Adapter {
+	if handler == nil {
 		return NoopAdapter
 	}
 
+	// NOTE(dlk): add constant values to include on every message here
+	ls := slog.New(handler.WithAttrs([]slog.Attr{
+		{Key: trails.LogKindKey, Value: trails.HTTPLogKind},
+	}))
+
 	return func(h http.Handler) http.Handler {
 		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			uri := r.URL.Path
-			q := r.URL.Query()
-			if val := q.Get("password"); val != "" {
-				q.Set("password", "xxxxxxx")
-			}
-
-			query := q.Encode()
-			if query != "" {
-				uri += "?" + q.Encode()
-			}
-
-			strs := []string{r.Method, uri}
-			val := r.Context().Value(IpAddrCtxKey)
-			if val != nil {
-				strs = append([]string{val.(string)}, strs...)
-			}
-
-			ls.Info(strings.Join(strs, " "), nil)
-			h.ServeHTTP(w, r)
+			start := time.Now()
+
+			writer := &requestLogger{ResponseWriter: w, status: http.StatusOK}
+			h.ServeHTTP(writer, r)
+
+			end := time.Since(start).Milliseconds()
+
+			rec := newRecord(writer, r)
+			rec.Duration = end
+
+			var msg string // NOTE(dlk): no message for now.
+
+			// NOTE(dlk): LogAttrs is not the standard method to use,
+			// but we know we only have a []slog.Attr to report,
+			// so let's leverage it.
+			//
+			// TODO(dlk): using r.Context() may not be appropriate.
+			// Unfortunately, outside of configuring a slog.Handler to pull
+			// values from a context.Context into slog.Attr,
+			// contexts appear to not be used by slog.
+			// This includes not watching for cancellations.
+			// The ranger.Context may be the best fit for this.
+			// We'll reassess when slog makes into a Go version.
+			ls.LogAttrs(r.Context(), slog.LevelInfo, msg, rec.attrs()...)
 		})
 	}
 }
+
+// A LogRequestRecord represents the fields that a LogRequest
+type LogRequestRecord struct {
+	BodySize       int    `json:"responseBodySize"`
+	Duration       int64  `json:"duration"`
+	Host           string `json:"host"`
+	ID             string `json:"id"`
+	IPAddr         string `json:"remoteAddr"`
+	Method         string `json:"method"`
+	Path           string `json:"path"`
+	Protocol       string `json:"protocol"`
+	Referrer       string `json:"referrer"`
+	ReqContentLen  int    `json:"requestContentLength"`
+	ReqContentType string `json:"requestContentType"`
+	Scheme         string `json:"scheme"`
+	Status         int    `json:"status"`
+	URI            string `json:"uri"`
+	UserAgent      string `json:"userAgent"`
```

Comment:
I don't see `timestamp` anywhere -- is that something slog outputs automagically?

Reply:
Revisiting the `slog` docs, it looks like `time` is provided automagically? So if that's the case forget my request to rename it to `timestamp` -- no need to mess with its default (I've just always preferred the explicit `timestamp` when its a datetime so that there isn't the ❓ of whether there's ever both `time` and `date` fields).

Reply:
It is a little strange they went with `time`, maybe it'll change before merging in to the std lib. I'll show in a forthcoming PR how to overwrite these default keys, so we can relitigate there if need be.